barretenberg/cpp/src/barretenberg/honk/flavor/ultra.hpp

```diff
@@ -407,6 +408,138 @@ class Ultra {
         FF gate_separation_challenge;
         FF target_sum;
     };
+
+    class Transcript : public BaseTranscript<FF> {
+      public:
+        uint32_t circuit_size;
+        uint32_t public_input_size;
+        uint32_t pub_inputs_offset;
+        std::vector<FF> public_inputs;
+        Commitment w_l_comm;
+        Commitment w_r_comm;
+        Commitment w_o_comm;
+        Commitment sorted_accum_comm;
+        Commitment w_4_comm;
+        Commitment z_perm_comm;
+        Commitment z_lookup_comm;
+        std::vector<barretenberg::Univariate<FF, MAX_RELATION_LENGTH>> sumcheck_univariates;
+        std::array<FF, NUM_ALL_ENTITIES> sumcheck_evaluations;
+        std::vector<Commitment> gemini_univariate_comms;
+        std::vector<FF> gemini_a_evals;
+        Commitment shplonk_q_comm;
+        Commitment kzg_w_comm;
+
+        Transcript(uint32_t circuit_size) { setUpStructure(circuit_size); }
+
+        Transcript(uint32_t circuit_size, const std::vector<uint8_t>& proof_data)
+        {
+            setUpStructureAndDeserialize(circuit_size, proof_data);
+        }
+
+      private:
+        template <typename T> inline static TranscriptObjectType convertTypeToEnum([[maybe_unused]] T _)
+        {
+            if constexpr (std::same_as<T, uint32_t>) {
+                return UInt32Obj;
+            } else if constexpr (std::same_as<T, FF>) {
+                return FieldElementObj;
+            } else if constexpr (std::same_as<T, Commitment>) {
+                return CommitmentObj;
+            } else if constexpr (std::same_as<T, Univariate<FF, MAX_RELATION_LENGTH>>) {
+                return SumcheckUnivariateObj;
+            } else if constexpr (std::same_as<T, std::array<FF, NUM_ALL_ENTITIES>>) {
+                return SumcheckEvalObj;
+            } else {
+                throw_or_abort("Received unknown type in convertTypeToEnum");
+            }
+        }
+
+        static std::vector<uint8_t> serializeObj(TranscriptObjectType enum_type, void* obj_ptr)
```

Comment:
I wonder if its possible to replace the void* usage with a type safe union instead, something like:

```cpp
#include <iostream>
#include <string>
#include <variant>

// Define a structure for each variant 
// (The code already has this defined, see uint32_t, FF, Commitment, ...)
struct VariantOne {
    int value;
};

struct VariantTwo {
    std::string message;
};

// Define the enum as a variant of all of your structures
using TranscriptObjectTypes = std::variant<VariantOne, VariantTwo>;

// Function to handle both variants
void handleVariant(const TranscriptObjectTypes& object_types) {
    std::visit([](auto&& arg) {
        using T = std::decay_t<decltype(arg)>;
        if constexpr (std::is_same_v<T, VariantOne>) {
            std::cout << "Handling VariantOne: " << arg.value << std::endl;
        } 
        else if constexpr (std::is_same_v<T, VariantTwo>) {
            std::cout << "Handling VariantTwo: " << arg.message << std::endl;
        }
    }, object_types);
}

int main() {
    TranscriptObjectTypes object_type = VariantOne{20694};
    handleVariant(object_type);
    
    object_type = VariantTwo{"Hello Rust!"};
    handleVariant(object_type);

    
    return 0;
}
```

Reply:
This may not be the best way to do it, just pointing out alternatives to void*

Reply:
forgot to respond - thanks for the suggestion! certainly a better solution than void*, but I think msgpack solves these problems

Reply:
can you elaborate how msgpack help this? I don't think msgpack should live at such a low level in the protocol